RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpMetricCalculator.cs

```diff
@@ -40,41 +40,43 @@ public CaDETClassMetrics CalculateClassMetrics(CaDETClass parsedClass)
         private double? GetTightClassCohesion(CaDETClass parsedClass)
         {
             int N = GetNumberOfMethodsDeclared(parsedClass);
+           
             double NP = (N * (N - 1)) / 2;
             if (NP == 0) return null;
 
-            return NumberOfDirectConnections(parsedClass) / NP;
+            return GetNumberOfDirectConnections(parsedClass) / NP;
         }
 
-        private int NumberOfDirectConnections(CaDETClass parsedClass)
+        private int GetNumberOfDirectConnections(CaDETClass parsedClass)
         {
-            int mapIndex = 1;
-            var map = new Dictionary<int, List<CaDETMember>>();
+            var ownAccessedFieldsAndAccessors = new Dictionary<string, List<CaDETMember>>();
 
             foreach (var method in parsedClass.Methods)
             {
-                map.Add(mapIndex, OwnAccessedFieldsAndAccessors(method));
-                mapIndex++;
+                ownAccessedFieldsAndAccessors.Add(method.Name, OwnAccessedFieldsAndAccessors(method));
             }
 
-            int countDirectConnections = 0;
+            return CountNumberOfDirectConnections(ownAccessedFieldsAndAccessors);
+        }
+
+        private static int CountNumberOfDirectConnections(Dictionary<string, List<CaDETMember>> ownAccessedFieldsAndAccessors)
+        {
+            int directConnections = 0;
 
-            foreach (var pair1 in map)
+            foreach (var ownAccessedFieldAndAccessor1 in ownAccessedFieldsAndAccessors)
```

Comment:
Having a suffix like "1" or "2" in your name is a code smell.  Unfortunately, I am not sure how to rename these iterators, which is a hint that something is faulty in the logic.

Instead of iterating over the Dictionary, can we iterate over the keys? Dictionaries are not really meant for iterating over. In general, this is reserved for lists and other Enumerables. I would refactor this code like so:

for(var i = 0; i < classMethods.length - 1; i++) //might be size instead of length, no intellisense to help me :)
{
  for(var j = 1; j < classMethods.length; j++)
  {
    var firstMethod = classMethods[i];
    var secondMethod = classMethods[j];
    if(firstMethod.GetAccessedOwnFieldsAndAccessors().Intersect(secondMethod.GetAccessedOwnFieldsAndAccessors()).Any()) directConnections++;
  }
}

Notice all the implications of the code above:
- I have moved the method GetAccessedOwnFieldsAndAccessors to the CaDETMember - this way we reduce feature envy and the parameter count.
- By doing this, I don't really need a Dictionary anymore, just the list of classMethods as the input to this function.
- I have also moved the word "Own" in the name of the method to (hopefully) make it a bit more clear what the meaning of this word is. It is still not ideal, though and requires the programmer to be somewhat familiar with the codebase.

With this transformation we should be able to actually remove the "GetNumberOfDirectConnections" and the TCC calculator can call the "CountMethodPairsThatShareAccessToAFieldOrAccessor" directly.

Final thought - this is a significant transformation of the code that I am suggesting and I might have poorly explained some part of it. Make sure to _meditate_ on both my comment and the broader implications of this transformation.

Reply:
Resolved

Reply:
Now we have two redundant methods, your CountOwnFieldAndAccessorAccessed(...parms) and my GetAccessedOwnFieldsAndAccessors() in CaDETmember.cs. 

![image](https://user-images.githubusercontent.com/49925421/97556670-0dcc7480-19da-11eb-85b3-bbb5927bbda5.png)

I suggest for the purpose of Clean Code, deleting your method, and use my in context GetAccessedOwnFieldsAndAccessors.Count.